File: pt6/1245.tree-diameter.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

class Solution {
    public int treeDiameter(int[][] edges) {
        if (edges.length == 0) {
            return 0;
        }

        int n = edges.length + 1;
        List<Integer>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int[] edge : edges) {
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
        }

        int endpoint = farthest(graph, 0)[0];
        return farthest(graph, endpoint)[1];
    }

    private int[] farthest(List<Integer>[] graph, int start) {
        Queue<int[]> queue = new ArrayDeque<>();
        boolean[] seen = new boolean[graph.length];
        queue.offer(new int[] {start, 0});
        seen[start] = true;

        int farNode = start;
        int farDist = 0;
        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            if (cur[1] > farDist) {
                farNode = cur[0];
                farDist = cur[1];
            }
            for (int next : graph[cur[0]]) {
                if (!seen[next]) {
                    seen[next] = true;
                    queue.offer(new int[] {next, cur[1] + 1});
                }
            }
        }

        return new int[] {farNode, farDist};
    }
}

/*
Explanation

In any tree, if we start from an arbitrary node and find the farthest node A,
A is an endpoint of the diameter. A second BFS from A gives the diameter
length.

The adjacency-list array is the natural Java representation for a tree with
node ids 0..n-1. BFS uses an ArrayDeque queue and a boolean visited array.

Edge cases: a single-node tree has diameter 0; a chain returns n - 1; a star
returns 2.

Time complexity: O(n), two BFS traversals.
Space complexity: O(n).
*/

File: Goldman Sachs/1245.tree-diameter.py
#
# @lc app=leetcode id=1245 lang=python3
#
# [1245] Tree Diameter
#
# https://leetcode.com/problems/tree-diameter/description/
#
# algorithms
# Medium (61.25%)
# Likes:    904
# Dislikes: 26
# Total Accepted:    57.4K
# Total Submissions: 93.7K
# Testcase Example:  "[[0,1],[0,2]]"
#
#
# The diameter of a tree is the number of edges in the longest path in
# that tree.
#
# There is an undirected tree of n nodes labeled from 0 to n - 1. You are
# given a 2D array edges where edges.length == n - 1 and edges[i] = [a_i,
# b_i] indicates that there is an undirected edge between nodes a_i and
# b_i in the tree.
#
# Return the diameter of the tree.
#
# Example 1:
#
# Input: edges = [[0,1],[0,2]]
# Output: 2
# Explanation: The longest path of the tree is the path 1 - 0 - 2.
#
# Example 2:
#
# Input: edges = [[0,1],[1,2],[2,3],[1,4],[4,5]]
# Output: 4
# Explanation: The longest path of the tree is the path 3 - 2 - 1 - 4 - 5.
#
# Constraints:
#
# n == edges.length + 1
#
# 1 <= n <= 10^4
#
# 0 <= a_i, b_i < n
#
# a_i != b_i
#
# @lc code=start
from typing import List
from collections import defaultdict, deque

class Solution:
    def treeDiameter(self, edges: List[List[int]]) -> int:
        """
        Interview explanation:
        Premium. Tree diameter = longest path (#edges). Two BFS: from any node
        find farthest u; from u find farthest v; distance u→v is diameter.

        Algorithm:
        - Build adj; BFS(0)→u; BFS(u)→(v, dist); return dist

        Complexity: O(n) time/space.
        """
        if not edges:
            return 0
        adj = defaultdict(list)
        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)

        def bfs(start: int):
            q = deque([(start, 0)])
            seen = {start}
            far_node, far_dist = start, 0
            while q:
                u, d = q.popleft()
                if d > far_dist:
                    far_dist, far_node = d, u
                for v in adj[u]:
                    if v not in seen:
                        seen.add(v)
                        q.append((v, d + 1))
            return far_node, far_dist

        u, _ = bfs(0)
        _, diameter = bfs(u)
        return diameter

    def treeDiameter_dfs(self, edges: List[List[int]]) -> int:
        """
        Interview explanation:
        Alternate: DFS/tree DP — for each node, track top-2 child heights;
        diameter = max over nodes of h1+h2.

        Algorithm:
        - Build adj; dfs returns height; update ans with sum of two largest child heights

        Complexity: O(n) time/space.
        """
        if not edges:
            return 0
        adj = defaultdict(list)
        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)
        ans = 0

        def dfs(u: int, parent: int) -> int:
            nonlocal ans
            top1 = top2 = 0
            for v in adj[u]:
                if v == parent:
                    continue
                h = dfs(v, u) + 1
                if h > top1:
                    top2, top1 = top1, h
                elif h > top2:
                    top2 = h
            ans = max(ans, top1 + top2)
            return top1

        dfs(0, -1)
        return ans
# @lc code=end
